doc/netlcc.1
.ad b
.TH netlcc 8 "25 June 1999" "CORE software" "CORE software"
.AT 3
.de sh
.br
.ne 5
.PP
\fB\\$1\fR
.PP
..
.PP
.SH NAME
.PP
.I netlcc
- convert 
.I netl
configuration files into 
.I netl
modules
.PP
.SH SYNOPSIS
.PP
.I netlcc
[gcc options] netl.conf
.PP
.SH DESCRIPTION
.PP
PLEASE NOTE: the 
.I netl
compiler is still somewhat experimental.  you
are encouraged to use it to generate 
.I netl
modules which you then
modify by hand.
.PP
.I netlcc
takes a 
.I netl
configuration file and converts it into a
.I netl
module, which can be loaded dynamically.  the advantage to a
.I netl
module is that it is compiled code and therefore theoretically
faster.  any .conf files are given to 
.BR netl (8)
to generate a .c file.  
this .c file is then compiled using 
.BR gcc (1)
, with any other options
that have been passed to the command line.  by default, a .so (shared
object file) is generated, which can be copied into the 
.I netl
modules
directory.
.PP
for more information on the 
.I netl
configuration format, see
.BR netl.conf (5)
.
.PP
for more information on 
.I netl
modules, see 
.BR netl_module (1)
.
.PP
for more information on 
.I netl
in general, see 
.BR netl (8)
.
.PP
.SH OPTIONS
.PP
.I netlcc
grabs two options, if seen, and passes the rest to gcc when it
reaches the compile stage.  therefore all the options you would expect to
work for a compiler such as -c or -O2 should work with 
.I netlcc
.
.PP
-generate-c
.PP
	generate c code and stop.  this is handy, if you want to hand
optimize the module generated from your configuration file.
.PP
-install
.PP
	once the module has been compiled successfully, this option will
direct 
.I netlcc
to install it in the correct place so that 
.BR netl (8)
can find it.  this option only works when compiling modules from
configuration files.
.PP
.SH SEE ALSO
.PP
.BR netl (8)
, 
.BR netl.conf (5)
, 
.BR netlcc (1)
, 
.BR netl_install (1)
,
.BR netl_module (1)
, 
.BR neta (1)
, 
.BR dcp (1)
and 
.BR xd (1)
.PP
.SH BUGS
.PP
there are almost certainly bugs, please report them to me.  my email
address is Graham Ollis <ollisg@wwa.com>, and the CORE home page is at
http://www.wwa.com/~ollisg/core.html, which should contain up to date
information on 
.B netl
.
.PP
.SH COPYING
.PP
Copyright 1996, 1997, 1999 Graham THE Ollis
.PP
This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.
.PP
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
Public License for more details.
.PP
You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
675 Mass Ave, Cambridge, MA 02139, USA.
.PP
.PP
.SH AUTHOR
.PP
Graham THE Ollis <ollisg@wwa.com>


HISTORY
version .95.2
-------------

 . mostly internal stuff.  added parse.c and grab.c modules, which contain
   code that used to be in the netl.c main module.  this is so that all the 
   network specific code is in grab.c, the true nightmare for porting is
   in this file... so if i can port grab.c to win32... then i'm done.

 . the \ character now works in the config file, if you want to have
   multiple line rules.  very unix-esq.

version .95
-----------

 . created man page documentation.

 . added `detect' config line, see the netl(8) manpage for details.
   this will reduce the amount of configeration you will have to do
   in the netl.conf file.

 . modified dist version of netl.conf so that it's more portable.
   in particular, i put in literal port numbers rather than using
   port names.

 . dcp.pl uses perl native Socket module instead of using net cat.
   `dcp.pl -o' uses the old method.

version .94
-----------

 . changed all the struct declerations to typedef declerations.  get off my
   back ok?  i've been programming perl for so long, forgot some of that
   strange C language. 

 . no longer depending on linux .h files for IP (and sub) headers.  this
   is for portability reasons.  if you are doing a port, make sure you get
   the endian type right!  there are now three global headers:
	global.h - general global stuff (endian integer conversion etc.)
	ether.h - ethernet specific data types, #defines
	ip.h - ip protocol specific data types, #defines
   logically, if your adding an interface (ATM, token ring, ppp, etc.)
   you'd want to start in ether.h.  if your adding an OS, look at global.h
   first.  if you want to add a protocol (IPX, etc.) start with ip.h and
   modify/add to these files.  if your trying to do all of this at once,
   god help you. 

 . fixed the clearipcache() function.  thus now, when you send a 
   netl:readconfig, netl will clear the ip cache before reading the file,
   avoiding all sorts of potential duplicates.  though you probably have
   an enourmous amount of memory like me and don't care... 

 . added some global options with ports to win32 in mind.  NO_SYSLOGD, if 
   defined in global.h will not even compile with the syslogd library stuff.  
   this is a good idea if your running from a non unix system.  note that 
   syslogd is really nice in that it partially compresses lines by indicating 
   repeated lines.  this won't save you if you get strobed, but will keep your 
   log file neat if some one ping -f's you from only one site.

 . added NO_TEEOUT option which is defined for unix.  the -o option should 
   only be necessary in win32 only.  even then, you should only compile with
   this option if your thinking of using it.  -o copies output to the specified
   file.

0.93 (and 0.92)
---------------

 a. -z option allows you to run netl in the foreground.  output is sent to
    stdout instead of syslogd.

 b. you can include configeration lines in the command line.  when you do
    this, the config file /etc/netl.conf will NOT be read.  note that (a)
    and (b) are meant to be used in concert with each other for testing
    and quick monitoring. 

 c. dcp is partially implemented.  netl will listen to port 47 udp
    (silently) if you `listen 47' in your config file.  when you send the
    message `netl:readconfig' netl is suppose to re read your config file. 
    dcp.pl is a perl script (you'll need net cat for this to work) which
    sends this message.  what is left to do for dcp?
	1. generic server side dcp interface libdcp.a file or somesuch.
	2. more solid C version of dcp client.

 d. netl is now officially under the GPL, see COPYING for more information.

 e. icmp codes are now included in some .h file somewhere rather than
    depending on your linux being overly up to date.

 f. neta no longer swaps source/destination ip/ethernet addresses
    inapropriately. 

 g. started work on hwpassive, which is no where near complete.

 h. tdr.pl is no included with netl distribution for convenience.  get
    your own version at: 

http://www.u.arizona.edu/~ollisg/core.html
	[ CORE ]

netl 0.91
---------

 minor updates

netl 0.90
---------

 initial release


netlcc.pl
#!/usr/bin/perl
# @(#)netlcc.pl (c) 1999 Graham THE Ollis
#===============================================================================
# front end for the netl c compiler.  this program will generate c code given
# a netl .conf file, then it will involk gcc to compile it.
#
#   Copyright (C) 1997 Graham THE Ollis <ollisg@wwa.com>
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
#===============================================================================

$netl = 'netl';
@netl_opts = ('-v-', '--generate-c');

$gcc = '/usr/bin/gcc';
@gcc_opts = ('-I/usr/local/lib/netl/include');

$install_dir = '/usr/local/lib/netl/filt';
	# note that we only install compiled .conf files.

$tmp_dir = '/tmp/.netl';
mkdir($tmp_dir, 0777);

@rest = (); @list = (); @mods = ();
for(@ARGV) {
	if(/\.conf$/) {

		($base = $in = $_) =~ s/\.conf$//;
		run($netl, @netl_opts, 
				'--file', $_, 
				'--output-name', "$tmp_dir/$base.c");
		push @list, $base;

	} elsif(/\.c$/) {
		push @mods, $_;
	} else {
		$target = 1 if /^-(c|S|E)$/;
		$no_gcc = 1 if /^-generate-c$/;
		$do_install = 1 if /^-install$/;
		if(/^--version$/) {
			print "netlcc version 1.02\n";
			run($gcc, '--version');
			exit;
		}
		push @rest, $_;
	}
}

if($no_gcc) {
	for(@list) {
		run('cp', "$tmp_dir/$_.c", '.');
		unlink "$tmp_dir/$_.c";
	}
	exit;
}


for(@mods) {
	($base = $_) =~ s/\.c//;
	@fred = ();
	@fred = ('-shared', '-o', "$base.so") unless $target;
	run($gcc, @gcc_opts, @rest, @fred, $_);
}

for(@list) {
	$base = $_;
	@fred = ();
	@fred = ('-shared', '-o', "$base.so") unless $target;
	run($gcc, @gcc_opts, 
		@rest, @fred,
		"$tmp_dir/$base.c");
	unlink "$tmp_dir/$base.c";
}

if($do_install) {
	for(@list) {
		if(-e "$base.so") {
			run('cp', "$base.so", $install_dir);
		}
	}
}

sub run {
	#print "run: @_\n";
	system(@_);
}
